docs/site/tutorials/authentication/Authentication-Tutorial.md

```diff
@@ -0,0 +1,974 @@
+      // don't copy over  token field 'iat' and 'exp', nor 'email' to user profile
+      userProfile = Object.assign(
+        {id: '', name: ''},
+        {id: decryptedToken.id, name: decryptedToken.name},
+      );
+    } catch (error) {
+      throw new HttpErrors.Unauthorized(
+        `Error verifying token: ${error.message}`,
+      );
+    }
+
+    return userProfile;
+  }
+
+  async generateToken(userProfile: UserProfile): Promise<string> {
+    if (!userProfile) {
+      throw new HttpErrors.Unauthorized(
+        'Error generating token: userProfile is null',
+      );
+    }
+
+    // Generate a JSON Web Token
+    let token: string;
+    try {
+      token = await signAsync(userProfile, this.jwtSecret, {
+        expiresIn: Number(this.jwtExpiresIn),
+      });
+    } catch (error) {
+      throw new HttpErrors.Unauthorized(`Error encoding token: ${error}`);
+    }
+
+    return token;
+  }
+}
+```
+
+`JWTService` generates or verifies JWT tokens using the `sign` and `verify`
+functions of [jsonwebtoken](https://www.npmjs.com/package/jsonwebtoken).
+
+It makes use of `jwtSecret` and `jwtExpiresIn` **string** values that are
+injected via the `TokenServiceBindings.TOKEN_SECRET` and the
+`TokenServiceBindings.TOKEN_EXPIRES_IN` binding keys respectively.
+
+The `async generateToken(userProfile: UserProfile): Promise<string>` function
+takes in a user profile of type
+[UserProfile](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/src/types.ts),
+generates a JWT token of type `string` using: the **user profile** as the
+payload, **jwtSecret** and **jwtExpiresIn**.
+
+The `async verifyToken(token: string): Promise<UserProfile>` function takes in a
+JWT token of type `string`, verifies the JWT token, and returns the payload of
+the token which is a user profile of type `UserProfile`.
+
+To bind the JWT `secret`, `expires in` values and the `JWTService` class to
+binding keys, we need to code the following in
+`loopback4-example-shopping/packages/shopping/src/application.ts`:
+
+```ts
+export class ShoppingApplication extends BootMixin(
+  ServiceMixin(RepositoryMixin(RestApplication)),
+) {
+  constructor(options?: ApplicationConfig) {
+    super(options);
+
+    // ...
+    this.setUpBindings();
+    // ...
+  }
+
+  setUpBindings(): void {
+    // ...
+
+    this.bind(TokenServiceBindings.TOKEN_SECRET).to(
+      TokenServiceConstants.TOKEN_SECRET_VALUE,
+    );
+
+    this.bind(TokenServiceBindings.TOKEN_EXPIRES_IN).to(
+      TokenServiceConstants.TOKEN_EXPIRES_IN_VALUE,
+    );
+
+    this.bind(TokenServiceBindings.TOKEN_SERVICE).toClass(JWTService);
+
+    // ...
+  }
+}
+```
+
+In the code above, `TOKEN_SECRET_VALUE` has a value of `'myjwts3cr3t'` and
+`TOKEN_EXPIRES_IN_VALUE` has a value of `'600'`.
+
+`JWTService` is used in two places within the application:
+`JWTAuthenticationStrategy` in
+`loopback4-example-shopping/packages/shopping/src/authentication-strategies/jwt-strategy.ts`,
```

Comment:
`JWTAuthenticationStrategy` in `loopback4-example-shopping/packages/shopping/src/authentication-strategies/jwt-strategy.ts` *to generate the token* and `UserController` in
`loopback4-example-shopping/packages/shopping/src/controllers/user.controller.ts` *to decode the token*.

Reply:
I will address them in a different PR. I waited until late tonight to be able to merge...my branch is always behind the master during the day...because travis takes so long to run.